pkg/logging/logging.go

```diff
@@ -85,6 +88,132 @@ var (
+func (e *Entry) WithField(key string, value interface{}) *Entry {
+	wrappedEntry := e.Entry.WithField(key, value)
+	return &Entry{
+		Entry:  wrappedEntry,
+		Logger: e.Logger,
+	}
+}
+
+// WithFields adds a map of fields to the Entry.
+func (e *Entry) WithFields(fields logrus.Fields) *Entry {
+	wrappedEntry := e.Entry.WithFields(fields)
+	return &Entry{
+		Entry:  wrappedEntry,
+		Logger: e.Logger,
+	}
+}
+
+// Warn logs a message at level Warn on the standard logger.
+func (e *Entry) Warn(args ...interface{}) {
+	e.Entry.Warn(args...)
+	metrics.IncWarningsMetric(e.Logger.subsystem)
+}
+
+// Warning logs a message at level Warn on the standard logger.
+func (e *Entry) Warning(args ...interface{}) {
+	e.Entry.Warning(args...)
+	metrics.IncWarningsMetric(e.Logger.subsystem)
+}
+
+// Error logs a message at level Error on the standard logger.
+func (e *Entry) Error(args ...interface{}) {
+	e.Entry.Error(args...)
+	metrics.IncErrorsMetric(e.Logger.subsystem)
+}
+
+// Warnf logs a message at level Warn on the standard logger.
+func (e *Entry) Warnf(format string, args ...interface{}) {
+	e.Entry.Warnf(format, args...)
+	metrics.IncWarningsMetric(e.subsystem)
+}
+
+// Warningf logs a message at level Warn on the standard logger.
+func (e *Entry) Warningf(format string, args ...interface{}) {
+	e.Entry.Warningf(format, args)
+	metrics.IncWarningsMetric(e.subsystem)
+}
+
+// Errorf logs a message at level Error on the standard logger.
+func (e *Entry) Errorf(format string, args ...interface{}) {
+	e.Entry.Errorf(format, args...)
+	metrics.IncErrorsMetric(e.subsystem)
+}
+
+// NewEntry returns a new instance of Entry.
+func NewEntry(logger *Logger) *Entry {
+	wrappedEntry := logrus.NewEntry(logger.Logger)
+	return &Entry{
+		Entry:  wrappedEntry,
+		Logger: logger,
+	}
+}
+
+// Logger is a wrapper for the Logrus Logger. It exposes its public methods, but
+// also adds a functionality of storing errors and warnings as Prometheus
+// metrics. For details about Logger in Logrus see:
+// https://godoc.org/github.com/sirupsen/logrus#Logger
+type Logger struct {
+	*logrus.Logger
+	subsystem string
+}
+
+// WithField adds a single field to the log entry.
+func (l *Logger) WithField(key string, value interface{}) *Entry {
+	wrappedEntry := l.Logger.WithField(key, value)
+	return &Entry{
+		Entry:  wrappedEntry,
+		Logger: l,
+	}
+}
+
+// WithFields adds a map of fields to the log entry.
+func (l *Logger) WithFields(fields logrus.Fields) *Entry {
+	wrappedEntry := l.Logger.WithFields(fields)
+	return &Entry{
+		Entry:  wrappedEntry,
+		Logger: l,
+	}
+}
+
+// WithSubsystem adds an information about subsystem to the log entry.
+func (l *Logger) WithSubsystem(subsystem string) *Entry {
+	entry := l.WithField(logfields.LogSubsys, subsystem)
+	entry.subsystem = subsystem
+	return entry
+}
+
+// NewLogger returns a new instance of Logger with default attributes.
+func NewLogger() *Logger {
+	wrappedLogger := logrus.New()
+	wrappedLogger.Formatter = setupFormatter()
```

Comment:
logrus accepts hooks. You can remove all the wrappers added and replace it with this:
```
// InitializeDefaultLogger returns a logrus Logger with a custom text formatter.
func InitializeDefaultLogger() *logrus.Logger {
	logger := logrus.New()
	logger.Formatter = setupFormatter()
	logger.Hooks.Add(metrics.MyHook)
	return logger
}
```

in metrics:
```
type MyHooker struct {
}

func (m MyHooker) Levels() []log.Level {
	// return all levels
	return nil
}

func (m MyHooker) Fire(e *log.Entry) error {
	LogLevelsCounter.WithLabelValues(e.Level.String()).Inc()
	return nil
}

var (
	MyHook = MyHooker{}
...
```